### ISSUE TITLE
--add-host example doesn't work as shown

### ISSUE THREAD

**Issue by kodedylf** [docker/cli#5558]
In the line I've referenced, a python3 web server is run to server the `hello` file.

However, as shown, the command will run in the foreground and the user will be unable to run the rest of the commands in the same terminal session. If the user presses ctrl-c to terminate python to be able to run the container, the example won't work since the web server isn't running.

My own solution was to run python3 in the background by adding `&` to the command. However then I needed to kill the web server after running the container. Getting the pid and killing python isn't very elegant. Hopefully you can come up with a better solution.

**Comment by thaJeztah** [comment]
Thanks for reporting!

Hm... yes, that's a good one; I know we tried to write up a minimal-as-possible example just to illustrate the concept, but indeed running things in the background can be slightly tricky, or at least to keep things "simple" as well as (where possible) working across platforms (mac, linux, windows).

I guess a "scrappy" workaround here could be to have the user open two separate shells; one to run the Python server and another one to run the docker commands to connect to it.


@neersighted @dvdksn any thoughts / ideas ?

**Comment by arpitgupta-it** [comment]
While using two separate terminal sessions is a simple and effective workaround, providing additional options like scripts or using Docker Compose can enhance the user experience.

**Comment by arpitgupta-it** [comment]
@kodedylf Rather than running a Python server directly on the host, you can run it inside a Docker container. This way, the server and the application are managed in a single Docker environment.

**Comment by thaJeztah** [comment]
Right, but the `host.docker.internal` feature was added for users to have an option to connect to a services that's _not_ running inside a container, so some service on the host. For the docker-compose case, the recommended approach would for the `curl-client` service to connect with the `python-server` service through the internal network (are least; "don't map ports that shouldn't be accessible publicly").

**Comment by thaJeztah** [comment]
I guess we could use `nc` to send a response (and terminate itself after); also means we don't need `python` (which may not be installed?). It may be a bit obscure though 🤔 

```bash
(echo -e 'HTTP/1.1 200 OK\r\n\r\nhello from host!') | nc -l 8000 & disown


ocker run -q --rm \
  --add-host host.docker.internal=host-gateway \
  curlimages/curl -s host.docker.internal:8000
```

```console
GET / HTTP/1.1
Host: host.docker.internal:8000
User-Agent: curl/8.13.0
Accept: */*

hello from host!
```

**Comment by dvdksn** [comment]
@thaJeztah I like it

**Comment by kodedylf** [comment]
There's not much difference between running `python` in the background (as I wrote I did in the first post) or `nc`.

If you want to be nice, you should clean up the background process and I don't know of an easy way to do that. Getting the pid of the background process and killing it is kind of messy.

**Comment by thaJeztah** [comment]
Yeah, that's where my thinking for `nc` came from here; it should only produce a response once, then terminate, so as long as the reader followed the example and made the connection, it would be gone afterwards.

**Comment by kodedylf** [comment]
I need to add a `-q 0` option on `nc` to get it to close the connection after sending the text. I'm on Ubuntu and I think it's the OpenBSD version of `nc`. I found [this](https://serverfault.com/questions/512722/how-to-automatically-close-netcat-connection-after-data-is-sent) which indicates that different versions of `nc` need different options to get them to close the connection.

When I ran it without `-q 0`, `curl` hang while waiting for the connection to get closed. And `nc` didn't terminate.

But with the `-q 0` option it works well for me.
